README.md
- [Getting started](#getting-started)
	- [Linking](#linking)
		- [Automatic](#automatic)
		- [Manual](#manual)
- [Usage](#usage)
	- [Provisioning a User](#provisioning-a-user)
	- [Managing a session](#managing-a-session)
		- [Events](#events)
	- [Example](#example)
## Getting started

```shell
$ npm install react-native-validic-session --save`
```
or use yarn

```shell
$ yarn add react-native-validic-session`
```

### Linking 

#### Automatic

```shell
react-native link react-native-validic-session
```

_For iOS users using Pods_
You still need to run `pod install` after running the above link command inside your `IOS` folder.

#### Manual

<details>
    <summary>iOS (via CocoaPods) RN >= 60</summary>

Add the following lines to your build targets in your `Podfile`

```ruby
pod 'React', :path => '../node_modules/react-native'

pod 'RNValidicSession', :path= => '../node_modules/react-native-validic-session'

```

Then run `pod install`

</details>


<details>
    <summary>iOS (without CocoaPods)</summary>

In XCode, in the project navigator:

- Right click _Libraries_
- Add Files to _[your project's name]_
- Go to `node_modules/react-native-validic-session/ios`
- Add the file `RNValidicHealthkit.xcodeproj`

In XCode, in the project navigator, select your project.

- Add the `libRNValidicHealthkit.a` from the _RNValidicHealthkit_ project to your project's _Build Phases ➜ Link Binary With Libraries_
- Click `.xcodeproj` file you added before in the project navigator and go the _Build Settings_ tab. Make sure _All_ is toggled on (instead of _Basic_).
- Look for _Header Search Paths_ and make sure it contains both `$(SRCROOT)/../react-native/React` and `$(SRCROOT)/../../React`
- Mark both as recursive (should be OK by default).


 In the General tab there will be a panel at the bottom of the screen labeled "Embedded Binaries". You should see the frameworks added here.

![Linked Frameworks](img/Linked_Frameworks.png)

Next copy 'copy-validicmobile.rb' to the `ios` folder of your project.

Go to the build settings of your target. In the "Build Phases" tab add a new build phase by clicking the plus button at the top and selecting "New Run Script Phase".

![Build Phase](img/Build_Phase.png)

In the new phase's text area paste in `ruby copy-validicmobile.rb`.

![Copy Script](img/Copy_Script.png)

</details>


<details>
	<summary>Android</summary>

1. Open up `android/app/src/main/java/[...]/MainApplication.java`
  - Add `import com.validic.mobile.rn.session.RNValidicMobileSessionPackage;` to the imports at the top of the file
  - Add `new RNValidicMobileSessionPackage()` to the list returned by the `getPackages()` method
2. Append the following lines to `android/settings.gradle`:
  	```
  	include ':react-native-validic-session'
  	project(':react-native-validic-session').projectDir = new File(rootProject.projectDir, 	'../node_modules/react-native-validic-session/android')
  	```
3. Insert the following lines inside the dependencies block in `android/app/build.gradle`:
  	```
      compile project(':react-native-validic-session')
  	```

</details>	  

## Usage

### Provisioning a User

The Validic Mobile library requires a Validic user. A Validic user can be provisioned using the [Validic API](http://validic.com/api#1649). The library requires a Validic user ID, user access token and organization ID. It is recommended to [provision the user](https://validic.com/contentblock/user-provisioning) on your server, the mobile app could then authenticate to the server and retrieve the Validic credentials.



```javascript
import ValidicSession from 'react-native-validic-session';
```

### Managing a session

The `ValidicSession` persists a user across the sdk and all pending record uploads. This data is persisted between
app launches but is deleted if the `ValidicSession.endSession` function is called.

A new Session is started by passing previously provisioned Validic user credentials to the `startSession` function.

```js
ValidicSession.startSession({user_id:USER_ID, org_id: ORGANIZATION_ID, user_token: USER_ACCESS_TOKEN});
```
The supplied credentials are not validated when the session is created. Records will fail to upload if invalid credentials have been provided. Credentials are stored persistently. To change credentials, call:

```js
ValidicSession.endSession();
```
or uninstall the app. When a new session is created, pending record uploads and subscriptons are reset. For this reason, a new session should only be created if a session it not already active.

The app can be informed of record upload success and failure by registering callbacks using:

#### Events
Listeners can be added to listen for success or failure of records submitted to the Validic Api. Registration of listeners is typically done during component initialization to catch events for record uploads which may occur upon app restart.

```js
ValidicSession.eventEmitter.addListener('validic:session:onsuccess', (record)=>{
	// the record was successfully uploaded to validic servers
	console.log("Successfully got record" + JSON.stringify(record));
});
ValidicSession.eventEmitter.addListener('validic:session:onerror', (error, record)=>{
	// an error was returned while submitting the record
	console.error("Got error: "+error + " for record: " + record.activity_id);
});
```

### Example

The following example starts a session if a session is not already active and registers session listeners.

```js

componentDidMount(){
ValidicSession.eventEmitter.addListener('validic:session:onsuccess', (record)=>{
	// the record was successfully uploaded to validic servers
	console.log("Successfully got record" + JSON.stringify(record));
});
ValidicSession.eventEmitter.addListener('validic:session:onerror', (error, record)=>{
	// an error was returned while submitting the record
	console.error("Got error: "+error + " for record: " + record.activity_id);
});

var user = {
    user_id: "replace with userid",
    user_token : "replace with access token",
    org_id : "replace with organization id"
};
// check if a persistent user already exists 
ValidicSession.getUser((response)=>{
	//if not, start a new session
	if(response.user == null){
		ValidicSession.startSession(user);
	}
});
}

...
//remove event listeners when the component is going to be unmounted 
componentWillUnmount(){
	ValidicSession.eventEmitter.removeAllListeners('validic:session:onsuccess');
	ValidicSession.eventEmitter.removeAllListeners('validic:session:onerror');
}

```

  

CHANGELOG.md
# Changelog

## 1.3.5 2020-04-05
- Bundled native SDKs upgraded to:
  - iOS: v1.12.11
  - Android: v1.12.15
  - 
## 1.3.4 3-08-2020
- Bundled native SDKs upgraded to:
  - iOS: v1.12.9
  - Android: v1.12.13
    
## 1.3.3. 2-24-2020
- Bundled native SDKs upgraded to:
  - iOS: v1.12.8
  - Android: v1.12.12
   
## 1.3.2 - 12-18-2020
- Bundled native SDKs upgraded to:
 - iOS: v1.12.6
 - Android v1.12.10

## 1.3.1 - 11-20-2020
- Bundled native SDKs upgraded to:
 - iOS: v1.12.5
 - Android v1.12.9
- Introduced Android 10, Android 11 and iOS 14 support

## 1.3.0 - 05-06-2020
- Bump native Android SDK release version 1.12.7
- Bump native iOS SDK release version 1.12.1

## 1.2.0 - 02-06-2020
- Use iOS Native SDK release version 1.11.5.
- Use Android Native SDK release version 1.12.5.
- Add WorkManager dependency for android

## 1.1.0 Date 10-22-2019
- Replace AndroidAsync with Retrofit and OkHttp for android 


index.js

import { NativeModules, NativeEventEmitter } from 'react-native';

/**
 * @namespace
 * @description
 * The ValidicSession allows for managing and observing a Validic user session with the Validic Mobile SDK. Starting a 
 * session is a required step before interacting with any Validic Mobile SDKs. 
 */
var ValidicSession = {
    /**
     * A Validic User e
     * @typedef User 
     * @memberof ValidicSession
     * @property {String} user_id      The Validic User Id to submit data to
     * @property {String} org_id       The Validic Organization the user_id belongs to
     * @property {String} user_token   The authroization token for the Validic user_id
     */

    /**
     * Start a Session with a Validic User 
     * @param {User} user 
     */
    startSession(user) {
        NativeModules.RNValidicMobileSession.startSession(user.user_id, user.org_id, user.user_token);
    },

    /**
     * End a Validic user's session. This will remove any pending work from the Validic Mobile SDK
     */
    endSession() {
        NativeModules.RNValidicMobileSession.endSession();
    },

    /**
     * @callback UserCallback - The callback containing the User credentials as added in {@function startSession}
     * @memberof ValidicSession
     * or nothing if the user is empty
     * @property {User} user - The Validic User 
     */

    /**
     * @param {UserCallback} callback
     * @memberof ValidicSession
     */
    getUser(callback) { NativeModules.RNValidicMobileSession.getUser(callback) },

    /**
     * @property {String} version - The native session library version
     */
    version: NativeModules.RNValidicMobileSession.SESSION_VERSION,

    /**
     * Events emitted by the native modules can be subscribed to here. Listeners can be registered to at any 
     * place in the app but should be unregistered when no longer needed
     */
    eventEmitter: new NativeEventEmitter(NativeModules.RNValidicMobileSession)

};

export default ValidicSession;



ios/Classes/RNValidicMobileSession.m
#import "RNValidicMobileSession.h"

@implementation RNValidicMobileSession{
    bool hasListeners;
}

RCT_EXPORT_MODULE()

- (dispatch_queue_t)methodQueue
{
    return dispatch_get_main_queue();
}

-(NSDictionary *)constantsToExport {
    return @{
             @"SESSION_VERSION" : [VLDSession libraryVersion]
             };
}

RCT_EXPORT_METHOD(startSession:(NSString *)userID organizationID:(NSString *)orgID accessToken:(NSString *)token) {
    VLDUser *user = [[VLDUser alloc] initWithValidicUserID:userID organizationID:orgID accessToken:token];
    [[VLDSession sharedInstance] startSessionWithUser:user];
}

RCT_EXPORT_METHOD(endSession) {
    [[VLDSession sharedInstance] endSession];
}

RCT_EXPORT_METHOD(submitRecord:(VLDRecord *)record) {
    [[VLDSession sharedInstance] submitRecord:record];
}

RCT_EXPORT_METHOD(getUser:(RCTResponseSenderBlock)callback) {
    VLDUser *user = [[VLDSession sharedInstance] user];
    
    if (user != nil) {
        callback( @[@{ @"user" : @{ @"user_id" : user.validicUserID,
                                    @"org_id" : user.organizationID,
                                    @"user_token" : user.accessToken
                                    }}]);
    } else {
        callback( @[@{ @"user" : [NSNull null] }]);
    }
    
}

- (NSArray<NSString *> *)supportedEvents {
    return @[@"validic:session:onsuccess", @"validic:session:onerror"];
}

- (void)startObserving {
    hasListeners = YES;
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didSubmitRecord:) name:kVLDRecordSubmittedNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(failedToSubmitRecord:) name:kVLDRecordSubmissionFailedNotification object:nil];
}

- (void)stopObserving {
    hasListeners = NO;
    [[NSNotificationCenter defaultCenter] removeObserver:self name:kVLDRecordSubmittedNotification object:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:kVLDRecordSubmissionFailedNotification object:nil];
}

- (void)didSubmitRecord:(NSNotification *)notification {
    VLDRecord *record = (VLDRecord *)notification.object;
    NSDictionary *recordDic = [record JSONObject];
    if(hasListeners){
        [self sendEventWithName:@"validic:session:onsuccess" body:recordDic];
    }
}

- (void)failedToSubmitRecord:(NSNotification *)notification {
    NSError *error = notification.userInfo[@"error"];
    VLDRecord *record = (VLDRecord *)notification.object;
    NSDictionary *dictionary = @{@"record": [record JSONObject], @"error": [error userInfo]};
    if(hasListeners){
        [self sendEventWithName:@"validic:session:onsuccess" body:dictionary];
    }
}

+(BOOL) requiresMainQueueSetup{
    return YES;
}


@end


android/src/main/java/com/validic/mobile/rn/session/Utils.java
package com.validic.mobile.rn.session;


import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;

/**
 * Created by griffinwilson on 2/21/18.
 */

public class Utils {

    static WritableMap convertJsonToMap(JSONObject jsonObject) throws JSONException {
        WritableMap map = Arguments.createMap();

        Iterator<String> iterator = jsonObject.keys();
        while (iterator.hasNext()) {
            String key = iterator.next();
            Object value = jsonObject.get(key);
            if (value instanceof JSONObject) {
                map.putMap(key, convertJsonToMap((JSONObject) value));
            } else if (value instanceof JSONArray) {
                map.putArray(key, convertJsonToArray((JSONArray) value));
            } else if (value instanceof Boolean) {
                map.putBoolean(key, (Boolean) value);
            } else if (value instanceof Integer) {
                map.putInt(key, (Integer) value);
            } else if (value instanceof Double) {
                map.putDouble(key, (Double) value);
            } else if (value instanceof String) {
                map.putString(key, (String) value);
            } else {
                map.putString(key, value.toString());
            }
        }
        return map;
    }

    static WritableArray convertJsonToArray(JSONArray jsonArray) throws JSONException {
        WritableArray array = Arguments.createArray();

        for (int i = 0; i < jsonArray.length(); i++) {
            Object value = jsonArray.get(i);
            if (value instanceof JSONObject) {
                array.pushMap(convertJsonToMap((JSONObject) value));
            } else if (value instanceof JSONArray) {
                array.pushArray(convertJsonToArray((JSONArray) value));
            } else if (value instanceof Boolean) {
                array.pushBoolean((Boolean) value);
            } else if (value instanceof Integer) {
                array.pushInt((Integer) value);
            } else if (value instanceof Double) {
                array.pushDouble((Double) value);
            } else if (value instanceof String) {
                array.pushString((String) value);
            } else {
                array.pushString(value.toString());
            }
        }
        return array;
    }

    static JSONObject convertMapToJson(ReadableMap readableMap) throws JSONException {
        JSONObject object = new JSONObject();
        ReadableMapKeySetIterator iterator = readableMap.keySetIterator();
        while (iterator.hasNextKey()) {
            String key = iterator.nextKey();
            switch (readableMap.getType(key)) {
                case Null:
                    object.put(key, JSONObject.NULL);
                    break;
                case Boolean:
                    object.put(key, readableMap.getBoolean(key));
                    break;
                case Number:
                    object.put(key, readableMap.getDouble(key));
                    break;
                case String:
                    object.put(key, readableMap.getString(key));
                    break;
                case Map:
                    object.put(key, convertMapToJson(readableMap.getMap(key)));
                    break;
                case Array:
                    object.put(key, convertArrayToJson(readableMap.getArray(key)));
                    break;
            }
        }
        return object;
    }

    static JSONArray convertArrayToJson(ReadableArray readableArray) throws JSONException {
        JSONArray array = new JSONArray();
        for (int i = 0; i < readableArray.size(); i++) {
            switch (readableArray.getType(i)) {
                case Null:
                    break;
                case Boolean:
                    array.put(readableArray.getBoolean(i));
                    break;
                case Number:
                    array.put(readableArray.getDouble(i));
                    break;
                case String:
                    array.put(readableArray.getString(i));
                    break;
                case Map:
                    array.put(convertMapToJson(readableArray.getMap(i)));
                    break;
                case Array:
                    array.put(convertArrayToJson(readableArray.getArray(i)));
                    break;
            }
        }
        return array;
    }

    static void sendEvent(ReactContext reactContext,
                          String eventName,
                          WritableMap params) {
        reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }

}


android/src/main/java/com/validic/mobile/rn/session/RNValidicMobileSharedPackage.java
package com.validic.mobile.rn.session;

import com.facebook.react.ReactPackage;
import com.facebook.react.bridge.JavaScriptModule;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.uimanager.ViewManager;
import com.validic.mobile.ValidicMobile;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class RNValidicMobileSharedPackage implements ReactPackage {
    @Override
    public List<NativeModule> createNativeModules(ReactApplicationContext reactContext) {
        ValidicMobile.getInstance().initialize(reactContext.getApplicationContext());
        return Arrays.<NativeModule>asList(new RNValidicMobileSessionModule(reactContext));
    }

    // Deprecated from RN 0.47
    public List<Class<? extends JavaScriptModule>> createJSModules() {
        return Collections.emptyList();
    }

    @Override
    public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
        return Collections.emptyList();
    }
}


android/src/main/java/com/validic/mobile/rn/session/RNValidicMobileSessionModule.java
package com.validic.mobile.rn.session;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;
import com.validic.common.GsonUtil;
import com.validic.common.ValidicLog;
import com.validic.mobile.Session;
import com.validic.mobile.SessionListener;
import com.validic.mobile.User;
import com.validic.mobile.ValidicMobile;
import com.validic.mobile.record.Record;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import javax.annotation.Nullable;

import static com.validic.mobile.rn.session.Utils.sendEvent;

/**
 * Created by griffinwilson on 2/21/18.
 */

public class RNValidicMobileSessionModule extends ReactContextBaseJavaModule implements SessionListener {

    public RNValidicMobileSessionModule(ReactApplicationContext reactContext) {
        super(reactContext);
        ValidicMobile.getInstance().initialize(reactContext.getApplicationContext());
        Session.getInstance().setSessionListener(this);
        ValidicLog.globalLoggingEnabled(BuildConfig.DEBUG);
    }

    @Override
    public String getName() {
        return "RNValidicMobileSession";
    }

    @ReactMethod
    public void startSession(String user, String org, String token) {
        Session.getInstance().startSessionWithUser(new User(org, user, token));
    }

    @ReactMethod
    public void endSession() {
        Session.getInstance().endSession();
    }

    @ReactMethod
    public void getUser(Callback userCallback) {
        User user = Session.getInstance().getUser();
        WritableMap callbackMap = Arguments.createMap();
        if (user != null) {
            WritableMap userMap = Arguments.createMap();
            userMap.putString("user_id", user.getValidicUserID());
            userMap.putString("org_id", user.getOrganizationID());
            userMap.putString("user_token", user.getAccessToken());
            callbackMap.putMap("user", userMap);
        } else {
            callbackMap.putMap("user", null);
        }
        userCallback.invoke(callbackMap);
    }

    @ReactMethod
    public void submitRecord(Record record) {
        Session.getInstance().submitRecord(record);
    }

    @Override
    public void didSubmitRecord(Record record) {
        try {
            JSONObject obj = new JSONObject(GsonUtil.getInstance().toJson(record));
            sendEvent(getReactApplicationContext(), "validic:session:onsuccess", Utils.convertJsonToMap(obj));
        } catch (JSONException e) {
            e.printStackTrace();
        }

    }

    @Override
    public void didFailToSubmitRecord(Record record, Error error) {
        try {
            JSONObject obj = new JSONObject(GsonUtil.getInstance().toJson(record));
            WritableMap map = Arguments.createMap();
            WritableMap rMap = Utils.convertJsonToMap(obj);
            map.putString("error", error.getMessage());
            map.putMap("record", rMap);
            sendEvent(getReactApplicationContext(), "validic:session:onerror", map);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    @Nullable
    @Override
    public Map<String, Object> getConstants() {
        Map<String, Object> constants = new HashMap<>();
        constants.put("SESSION_VERSION", Session.libraryVersion());
        return constants;
    }
}